File: vb/HtmlReport.vb
Imports System.Collections.Generic
Imports System.IO

Namespace ReportGeneratorVB
    Public Class HtmlReport
        Public Shared Sub report(ByVal output As StringWriter, ByVal machines As IList(Of Machine), ByVal robot As Robot)
            output.Write("<h1>FACTORY REPORT</h1>" & vbLf)

            Dim line As IEnumerator(Of Machine) = machines.GetEnumerator()
            output.Write("<ul>" & vbLf)
            While line.MoveNext()
                Dim machine As Machine = line.Current
                output.Write("<li>")
                output.Write("Machine " + machine.Name())

                If machine.Bin() IsNot Nothing Then
                    output.Write(" bin=" + machine.Bin())
                End If

                output.Write("</li>" & vbLf)
            End While
            output.Write("</ul>" & vbLf)

            output.Write("<p>")
            output.Write("Robot")
            If robot.Location() IsNot Nothing Then
                output.Write(" location=" + robot.Location().Name())
            End If

            If robot.Bin() IsNot Nothing Then
                output.Write(" bin=" + robot.Bin())
            End If

            output.Write("</p>" & vbLf)

            output.Write("<hr>" & vbLf)
        End Sub
    End Class
End Namespace

File: csharp/HtmlReport.cs
using System.Collections.Generic;
using System.IO;

namespace ReportGeneratorCS
{
    public class HtmlReport
    {
        public static void report(StringWriter output, IList<Machine> machines, Robot robot)
        {
            output.Write("<h1>FACTORY REPORT</h1>\n");

            IEnumerator<Machine> line = machines.GetEnumerator();
            output.Write("<ul>\n");
            while (line.MoveNext())
            {
                Machine machine = line.Current;
                output.Write("<li>");
                output.Write("Machine " + machine.Name());

                if (machine.Bin() != null)
                    output.Write(" bin=" + machine.Bin());

                output.Write("</li>\n");
            }
            output.Write("</ul>\n");

            output.Write("<p>");
            output.Write("Robot");
            if (robot.Location() != null)
                output.Write(" location=" + robot.Location().Name());

            if (robot.Bin() != null)
                output.Write(" bin=" + robot.Bin());

            output.Write("</p>\n");

            output.Write("<hr>\n");
        }
    }
}